File: go/2021/march/coin-change.go
package march

func coinChange(coins []int, amount int) int {
	max := amount + 1
	var dp = make([]int, max)
	for i := 0; i < max; i++ {
		dp[i] = max
	}
	dp[0] = 0
	for i := 1; i <= amount; i++ {
		for _, c := range coins {
			if c <= i {
				dp[i] = min(dp[i], dp[i-c]+1)
			}
		}
	}
	if dp[amount] > amount {
		return -1
	}
	return dp[amount]
}

func min(a, b int) int {
	if a > b {
		return b
	}
	return a
}

File: archive/cpp/2021/mar/coin-change.cc
#include <vector>

using std::vector;
class CoinChange
{
public:
    int coinChange(vector<int> &coins, int amount)
    {
        int max = amount + 1;
        vector<int> dp(max, max);
        dp[0] = 0;
        for (int i = 1; i <= amount; i++)
        {
            for (auto c : coins)
            {
                if (c <= i)
                {
                    dp[i] = min(dp[i], dp[i - c] + 1);
                }
            }
        }
        return dp[amount] > amount ? -1 : dp[amount];
    }
};

int min(int a, int b)
{
    return a > b ? b : a;
}